### PR TITLE
Azure Monitor: Add support for multiple template variables in resource picker

### LINKED ISSUE #43618: Azure Monitor DS: Logs resource picker does not have support for variables.
**What happened**:
On a dashboard with variables if a panel is created with Azure Monitor datasource and Logs as the service, when a variable is selected as the resource, Grafana crashes and throws the error "PathNotFound": 
image

After some inspection, it seems as though it is because the resource path is not being properly populated, it is rather just injecting the resource name, not a full path: 
image

**What you expected to happen**:

For the same resource selected above, except this time not through variables, this is what the user should see: 
image

and what the resource path should look like
image

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a dashboard with variables using the Azure Monitor data source. 
2. Add a panel with the data source as Azure Monitor and the service as Logs
3. Select any of the variables in the dashboard as the resource path
4. Add a simple query for grafana to run.

**Anything else we need to know?**:

**Environment**: 
- Grafana version: 8.4.0-pre
- Data source type & version: Azure Monitor
- OS Grafana is installed on: Windows
- User OS & Browser: Windows & Edge
- Grafana plugins:
- Others:

### LINKED ISSUE #46057: Azure Monitor: Add support for Template Variables in the Resource Picker
In one pr or multiple prs behind a feature flag of some sort:
 - [x] Replace existing selection row in Resource Picker with an input field
 - [x] Ensure when a selects something with the picker that the input field updates
 - [x] Ensure that when a user types in the input field, the picker updates 
 - [X] Ensure that multiple template variables are properly interpolated by the input field
 - [x] Ensure that if the user types something in the input field that can not be recognized by the picker that we handle this expected use-case gracefully with meaningful error messages that do not block the query. 

Consider running plan or WIP past UX rep or feedback session to ensure it's workable.

### PR DESCRIPTION
**Pull request by sarahzinger**
**What this PR does / why we need it**:
Screen Shot 2022-03-23 at 5 24 54 PM
Screen Shot 2022-03-23 at 5 24 38 PM

- removes the old template variable selector from the resource picker as it was pretty confusing and you could only use one template variable at a time
- adds a new input field where a user can type in the exact resource uri they want regardless of what's happening with the resource picker, including template variables

**Which issue(s) this PR fixes**:

Fixes #46057
Fixes #43618

### CODE REVIEW

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/styles.ts [hunk 1]
```diff
@@ -64,6 +64,10 @@ const getStyles = (theme: GrafanaTheme2) => ({
     textOverflow: 'ellipsis',
     whiteSpace: 'nowrap',
   }),
+
+  nestedRowCheckbox: css({
+    zIndex: 0,
```
Comment (sarahzinger) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/styles.ts:69]: unrelated css fix, was bothering me

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/NestedRows.tsx [hunk 2]
```diff
@@ -215,7 +215,13 @@ const NestedEntry: React.FC<NestedEntryProps> = ({
 
       {isSelectable && (
         <>
-          <Checkbox id={checkboxId} onChange={handleSelectedChanged} disabled={isDisabled} value={isSelected} />
+          <Checkbox
+            id={checkboxId}
+            onChange={handleSelectedChanged}
+            disabled={isDisabled}
+            value={isSelected}
+            className={styles.nestedRowCheckbox}
```
Comment (sarahzinger) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/NestedRows.tsx:223]: (put this in a separate pr here: #46665 will rebase after i merge it)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts [hunk 3]
```diff
@@ -68,14 +68,15 @@ export default class ResourcePickerData extends DataSourceWithBackend<AzureMonit
 
     return resources.map((subscription) => ({
       name: subscription.subscriptionName,
-      id: subscription.subscriptionId,
+      id: `/subscriptions/${subscription.subscriptionId}`,
```
Comment (sarahzinger) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:71]: (put this change in a separate pr here: #46665 will rebase after i merge it)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts [hunk 4]
```diff
@@ -68,14 +68,15 @@ export default class ResourcePickerData extends DataSourceWithBackend<AzureMonit
...
       typeLabel: 'Subscription',
       type: ResourceRowType.Subscription,
       children: [],
     }));
   }
 
-  async getResourceGroupsBySubscriptionId(subscriptionId: string) {
+  async getResourceGroupsBySubscriptionId(subscriptionURI: string) {
+    const subscriptionId = subscriptionURI.includes('/subscriptions/') ? subscriptionURI.slice(15) : subscriptionURI;
```
Comment (sarahzinger) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts:79]: (put this change in a separate pr here: #46665 will rebase after i merge it)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx [hunk 5]
```diff
@@ -150,33 +138,48 @@ const ResourcePicker = ({
...
+                      <a
+                        href="https://docs.microsoft.com/en-us/azure/azure-monitor/logs/log-standard-columns#_resourceid"
+                        rel="noopener noreferrer"
+                        target="_blank"
+                      >
+                        resource id (also referred to as a resource uri)
+                      </a>{' '}
+                      below. Grafana supports template variable interpolation like so:
+                    </p>
+                    <p>/subscriptions/$subId</p>
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:164]: What about "You can manually edit the [resource ID] or use template variables" (shorter but explaining everything?). If we need more info, we can add that to the docs and add a link here.

**Review by andresmgot — COMMENTED** [review]
A bit of UX feedback, seeing this, it seems that the resource URI is too prominent (while it's kind of an advanced feature). Maybe we can leave the selection as it was before and add a new collapsible section with the advanced feature. Something like:

Screenshot from 2022-03-17 13-04-58

With this:
 - People that are just interested in the same UX than the Azure Portal, don't need to be aware of the URI setup.
 - This URI input is explicitly disabled unless the user clicks on the "edit" button, this is to avoid accidental changes.
 - If the user edits this field, we can remove the "Selection" section and collapse all resources from the resource picker, that way we avoid buggy UI behaviors (like only showing the selection in the picker if the resource is loaded on the screen).

Also I notice that the resource pre-selection (when you re-open the picker) is not working but it may be fixed if you follow the approach above.

**Review by andresmgot — COMMENTED** [review]
Neat solution!

Another couple of minor things:

 - Can we add a bit more of margin between the "Selection" block and the Advanced one? So it's equally separated if the Selection does not exist:

Screenshot from 2022-03-24 10-38-51
Screenshot from 2022-03-24 10-37-53

- I noticed that if a template variable is used, we are no longer able to resolve the subscription name (and its ID is shown instead). Can we fix that?

Screenshot from 2022-03-24 11-00-24

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx [hunk 6]
```diff
@@ -125,28 +120,30 @@ describe('AzureMonitor ResourcePicker', () => {
     expect(onApply).toBeCalledWith('/subscriptions/def-123');
   });
 
-  it('should call onApply with a template variable when a user selects it', async () => {
```
[the remark this answers is not in the document]
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx:128]: I'm not sure the current template variables ui makes sense if you're using multiple template variables though? If you had a template variable it would still show in the advanced section?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx [hunk 7]
```diff
@@ -170,13 +153,53 @@ const ResourcePicker = ({
...
+                  >
+                    <Icon name="info-circle" />
+                  </Tooltip>
+                </h6>
+              </Label>
+              <Input
+                id={`input-${internalSelectedURI}`}
+                value={internalSelectedURI}
+                onChange={(event) => setInternalSelectedURI(event.currentTarget.value)}
+                placeholder="ex: /subscriptions/$subId"
```
[the remark this answers is not in the document]
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:192]: Hmm. We can? My thinking is though that if you're on the edit panel page and you open the resource picker your intention is to edit the panel. Seems odd to then make someone have to click again to edit. If you do mistype something in there you can reselect things in the resource picker and it will overwrite whatever you have in there or you can click cancel?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx:128]: no, this would be for the previously supported case of using just one template variable (and that single template variable is what it would be shown in the Advanced section). Does that make sense?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:192]: yeah, I was thinking on users clicking on "Advanced" out of curiosity and then editing by mistake the field but it's not something super worrying, up to you of whatever you decide.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx:128]: If they previously had just one template variable it would show in the advanced section. Is that not enough? I guess it doesn't hurt to bring it back! Can always remove it later if we think the UI is too cluttered. I guess I can see an argument that if you did have one template variable it might be annoying to type and so you'd rather just click on it.

[the remark this answers is not in the document]
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx:128]: eh ok. So I did experiment with bringing this back but I landed on feeling like it was more confusing than helpful. When we list the possible template variables to choose from, we make it seem like they are all valid options, when they might not be. If you want to use multiple template variable it can seem like you should be able to select multiple variables by clicking many checkboxes but this is not possible. 

I think your suggestion to keep advanced auto expanded when it detects that a user has a template variable will make it clear how to use it:
Here is what a user should see if they are currently using the workspaces() template variable for example
Screen Shot 2022-03-24 at 9 40 37 PM
Screen Shot 2022-03-24 at 9 40 44 PM

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (sarahzinger) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.tsx:191]: Since nothing updates until you click apply I think it's already safeguarded, but if folks make bugs saying they are confused maybe we can reopen!

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ResourcePicker/ResourcePicker.test.tsx:128]: fair enough! we lose the ability of listing the variables but it's clear / easy enough. Thanks for checking!

**Review by andresmgot — APPROVED** [review]
LGTM, good job!
